# Report
# Project Name - WhatTheRec

## Team Members: 
### Shiva Kumar Pentyala, Sai Kiran

## Introduction and Problem Statement
In the present scenario, a plethora of events is being held daily at different places and different times. Hence, it is difficult for a user to keep track of all the events and choose between various events to attend. The problem increases because the users do not know which events would be interesting and relevant to them. Hence, it is important that an event sharing social network uses recommenders to suggest relevant and interesting events to its users which help them choose between various events. While recommending events, another problem that arises is that of unseen data(events). We cannot say, confidently, which users will attend what events, especially in the case of new users. This is called the cold start problem. There are many Event Based Social Network (EBSN) websites like meetup.com, eventbrite.com etc., which could deploy this algorithm in order to successfully suggest interesting and relevant events to its users.

## Related Work
There has been a lot of work already done including hybrid approaches in event recommendation by Minkov et al. [1]. This paper demonstrates collaborative filtering on a dataset of academic events. Our implementation of the recommender system follows [2], which also utilizes other signals such as a group of users, location, and temporal preferences. Khrouf and Troncy [3] for recommending music related events, utilizes category information about different artists from a well-known source. But this approach fails when we do not have this information about every event, as events can be across different domains. Our implementation takes into account RSVPs, group information and other contextual data. Recent works [4] have shown that pure matrix factorization (based only on user-event interactions) performs poorly on EBSN data in comparison to other methods due to high level of the sparsity of these datasets. As per experiments carried out, state-of-the-art matrix factorization algorithms did not perform better than simple collaborative filtering algorithms such as user-based k-NN. Thus, our focus is on considering the explicit features than the latent ones.

## Approach & Methods

Data Acquisition - Initially we fetched the data from meetup.com using their public APIs. As we started developing, we realized that the data was quite less as it is super sparse in nature. And, when we again started getting data, they changed their APIs and a lot of them were broken. So, we decided to contact the authors of this paper and got the raw data from them. We cleaned and parsed the data and used that for our project. We used data from Jan 1, 2010 - Jan 2014. 
The data contains following information:

1. Events: event id, description of the event, event group, users who RSVP'd this event.
2. Members: member id, latitude, longitude
3. Groups: events organized by the group, members involved in the group

We, then, divided the data into timestamps of 6 months. For example, one timestamp was July 1, 2010. The first 6 months were used for training and the last 6 months for testing. We extracted the following features to recommend events: time, location, description and group frequency of the events. 

a. Content-based - We formed a model of a user using the description of the past events he/she had attended. Then, we found the similarity of potential events with the user model. 

b. Location-based - We fit a gaussian distribution to the location of the past events of a user. Then, we found the probability of new event according to the curve. 

c. Group-Frequency - The intuition here is that the likelihood of the target user attending an event depends on the number of events this user attended in the group that the event belongs to. 

d. Time-Aware - Another important factor that affects users' decision on attending an event is when the event occurs. We capture this intuition by assuming that users that attended events in the past on certain days of the week and at certain hours of the day will likely attend events with a similar temporal profile in the future.

Now, that we have built a model of the data, we then computed similarity scores for each member with the events that they had rsvp'd. Events in the first half were used to train the model and the events in the second half were used to test it. Note that we knew the events which the members had rsvp'd for the both the halves. So, the rsvp's for the second half were used to evaluate the model.

Now, you can run our recommender code below. Type the name of city you want to run the recommender on. 

Note: We have data only from Chicago, San Jose and Phoenix. You can also change the number of members you want to train on. Additionally, you can use a list of ML algorithms as well.

<b>Note: Please do not run this code outside the current directory. The code in this notebook calls some python code that are present in the subdirectories.</b>

Insert Parameters Below:

In [1]:
import os
os.chdir('src')
city = 'LCHICAGO' # or 'LSAN JOSE', or 'LPHOENIX'
algolist = ['rf'] #'svm', 'nb', 'mlp'
number_of_members = 50 #100, 150 (not more than this, otherwise system would take too much memory)


In [2]:
from preprocessing import *
import argparse
from partition import *
from content.content_recommender import ContentRecommender
from temporal.time_recommender import TimeRecommender
from location.location_recommender import LocationRecommender
from group_frequency.grp_freq_recommender import GrpFreqRecommender
from hybrid.learning_to_rank import LearningToRank
import datetime
import time

#number of seconds in 6 months
train_data_interval = ((364 / 2) * 24 * 60 * 60)


Training and Testing using Content Features:

In [3]:

def content_classifier(training_repo, test_repo, timestamp, simscores, test_members):

    training_events_dict = training_repo['members_events']
    potential_events = list(test_repo['events_info'].keys())

    contentRecommender = ContentRecommender()
    contentRecommender.train(training_events_dict, training_repo)
    test_events_vec = contentRecommender.get_test_events_wth_description(test_repo, potential_events)

    #TEST FOR BEST USERS
    for member in test_members:
         contentRecommender.test(member, potential_events, test_events_vec, simscores)


Training and Testing using Time Features

In [4]:
def time_classifier(training_repo, test_repo, timestamp, simscores, test_members):

    training_events_dict = training_repo['members_events']
    potential_events = list(test_repo['events_info'].keys())

    timeRecommender = TimeRecommender()
    timeRecommender.train(training_events_dict, training_repo)
    test_events_vec = timeRecommender.get_test_event_vecs_with_time(test_repo, potential_events)

    #TEST FOR BEST USERS
    for member in test_members:
         timeRecommender.test(member, potential_events, test_events_vec, simscores)


Training and Testing using Location feature

In [5]:
def loc_classifier(training_repo, test_repo, timestamp, simscores, test_members):
    training_events_dict = training_repo['members_events']
    potential_events = list(test_repo['events_info'].keys())

    locationRecommender = LocationRecommender()
    locationRecommender.train(training_events_dict, training_repo)

    #TEST FOR BEST USERS
    for member in test_members:
        locationRecommender.test(member, potential_events, test_repo, simscores)
        

Training and Testing using Group Frequency feature:

In [6]:

def grp_freq_classifier(training_repo, test_repo, timestamp, simscores, test_members):
    training_events_dict = training_repo['members_events']
    potential_events = list(test_repo['events_info'].keys())

    grp_freq_recommender = GrpFreqRecommender()
    grp_freq_recommender.train(training_events_dict, training_repo)
    
    for member in test_members:
         grp_freq_recommender.test(member, potential_events, test_repo, simscores)

    


Now, extract the data from csv file into python dictionaries. local_crawler.py performs this task.

In [7]:
def check_and_run_local_crawler():
    
    if os.path.isdir("../crawler/cities/LCHICAGO") and os.path.isdir("../crawler/cities/LSAN JOSE")\
            and os.path.isdir("../crawler/cities/LPHOENIX"):
        if len(os.listdir("../crawler/cities/LCHICAGO")) >= 5 and len(os.listdir("../crawler/cities/LSAN JOSE"))>=5\
                and len(os.listdir("../crawler/cities/LPHOENIX")) >= 5:
            return
        
    # Run the local_crawler.py file    
    os.chdir("../crawler")
    os.system("python local_crawler.py")
    os.chdir("../src")

This function generates a list of best users based on the number of RSVP's.

In [8]:
def run_script(number_of_members):
    os.chdir("scripts")
    os.system("python script.py --number " + str(number_of_members))
    os.chdir("..")

In [9]:
def main():
    
    check_and_run_local_crawler()
    print "Building best user database ..."
    run_script(number_of_members)
    print "Best users extracted."

    
    group_members, group_events, event_group = load_groups("../crawler/cities/" + city + "/group_members.json",
                                                            "../crawler/cities/" + city + "/group_events.json")
    events_info = load_events("../crawler/cities/" + city + "/events_info.json")
    members_info = load_members("../crawler/cities/" + city + "/members_info.json")
    member_events = load_rsvps("../crawler/cities/" + city + "/rsvp_events.json")

    repo = dict()
    repo['group_events'] = group_events
    repo['group_members'] = group_members
    repo['events_info'] = events_info
    repo['members_info'] = members_info
    repo['members_events'] = member_events
    repo['event_group'] = event_group
    
    #simscores_across_features is a dictionary to store similarity score obtained for each feature
    #for each member and for a given event. For example in case of content classifer we will
    #access the similarity score as follows: simscores['content_classifier'][member_id][event_id].
    #We will pass only a specific subdictionary (Ex: simscores['content_classifier']) to the
    #classifier functions, which will work on them and populate them.
    
    simscores_across_features = defaultdict(lambda :defaultdict(lambda :defaultdict(lambda :0)))
    hybrid_simscores = defaultdict(lambda :defaultdict(lambda :0))

    start_time = 1262304000 # 1st Jan 2010
    end_time = 1388534400 # 1st Jan 2014
    timestamps = get_timestamps(start_time, end_time)
    timestamps = sorted(timestamps, reverse=True)
    count_partition = 1

    f_temp = open('temp_result.txt', 'w+')
    f_temp.write("Using classification algorithms : " + str(algolist) + " and number of members as : " +\
                 str(number_of_members) + "\n")

    for t in timestamps:
        start_time = t - train_data_interval
        end_time = t + train_data_interval
        test_members = []
        f = open("scripts/"+city + "_best_users_" + str(start_time) + "_" + str(end_time) + ".txt", "r")
        for users in f:
            test_members.extend(users.split())
        f.close()
        test_members = test_members[:number_of_members]
        print "Partition at timestamp ", datetime.datetime.fromtimestamp(t), " are : "
        training_repo, test_repo = get_partitioned_repo_wrapper(t, repo)
        print "Partitioned Repo retrieved for timestamp : ", datetime.datetime.fromtimestamp(t)

        training_members = set(training_repo['members_events'].keys())
        test_members =  training_members.intersection(set(test_members))
        test_members = list(test_members)
        
        #Calling content based classifer train and test functions from here. Pass the repo
        #as an argument to these functions.
        start = time.clock()
        print "Starting Content Classifier"
        content_classifier(training_repo, test_repo, t, simscores_across_features['content_classifier'],\
                           test_members)
        print "Completed Content Classifier in ", time.clock() - start, " seconds"

        start = time.clock()
        print "Starting Time Classifier"
        time_classifier(training_repo, test_repo, t, simscores_across_features['time_classifier'], test_members)
        print "Completed Time Classifier in ", time.clock() - start, " seconds"

        start = time.clock()
        print "Starting Location Classifier"
        loc_classifier(training_repo, test_repo, t, simscores_across_features['location_classifier'], test_members)
        print "Completed Location Classifier in ", time.clock() - start, " seconds"

        start = time.clock()
        print "Starting Group Frequency Classifier"
        grp_freq_classifier(training_repo, test_repo, t, simscores_across_features['grp_freq_classifier'],\
                            test_members)
        print "Completed Group Frequency Classifier in ", time.clock() - start, " seconds"

        f_temp.write("============== Starting classification for partition : " +  str(count_partition) + \
                     " ===================\n")
        print "============== Starting classification for partition : " +  str(count_partition) + \
        " ==================="
        learningToRank = LearningToRank()
        learningToRank.learning(simscores_across_features, test_repo["events_info"].keys(), \
                                test_repo["members_events"], test_members, f_temp, algolist, \
                                number_of_members, count_partition)
        f_temp.write("============== Completed classification for partition : " +  str(count_partition) + \
                     " ===================\n")
        print "============== Completed classification for partition : " + str(count_partition) + \
        " ==================="
        
        count_partition += 1
    
    f_temp.close()

if __name__ == "__main__":
    main()


Building best user database ...
Best users extracted.
Partition at timestamp  2013-06-27 19:00:00  are : 
Partitioned Repo retrieved for timestamp :  2013-06-27 19:00:00
Starting Content Classifier
Completed Content Classifier in  183.34123  seconds
Starting Time Classifier
Completed Time Classifier in  6.515163  seconds
Starting Location Classifier
Completed Location Classifier in  203.89219  seconds
Starting Group Frequency Classifier
Completed Group Frequency Classifier in  32.044853  seconds
============== Starting classification for partition : 1 ===================
RF -> Precision : 0.695475638051 Recall : 0.872634643377 F-measure : 0.774047772757
============== Completed classification for partition : 1 ===================
Partition at timestamp  2012-12-27 18:00:00  are : 
Partitioned Repo retrieved for timestamp :  2012-12-27 18:00:00
Starting Content Classifier
Completed Content Classifier in  124.92545  seconds
Starting Time Classifier
Completed Time Classifier in  4.271634 

## Evaluation

We divided an year's data into 2 slots. We built the model of the user from the rsvp'd events in the first slot, and then for each of the event in the next slot, we performed the recommendation. Then, we did a k fold setup to evaluate recommendations. For 80% of the users, we trained the classifier and then calculated precision, recall and F-score for the remaining 20% users.

For a new user (who didn't attend any event in the fist slot), we recommend the event which occurs closest to the user's location.

We chose three cities in the USA for performing the evaluation, namely, Chicago, Phoenix and San Jose. There are 2 main reasons why we selected these 3 cities:- 

(1) They are among the most populous cities in the USA, which indicate large event activity 

(2) The cities are located in different states, thus representing some degree of cultural diversity. 

Following are the results for these cities. The results were captured on the basis of partition number and number of members on which evaluation was done. 

<h3>Results for Chicago:-</h3>
<table>
<tbody>
<tr><th>Partition Number</th><th>Number of members</th><th>Precision</th><th>Recall</th><th>F-measure</th></tr>
<tr>
<td rowspan="3">1</td>
<td>50</td>
<td>0.695475638051</td>
<td>0.872634643377</td>
<td>0.774047772757</td>
</tr>
<tr>
<td>100</td>
<td>0.700596056855</td>
<td>0.648556876061</td>
<td>0.673572845493</td>
</tr>
<tr>
<td>150</td>
<td>0.537199542159</td>
<td>0.529721595184</td>
<td>0.533434362569</td>
</tr>
<tr>
<td rowspan="3">
 2</td>
<td>50</td>
<td>0.660746003552</td>
<td>0.749244712991</td>
<td>0.702218027371</td>
</tr>
<tr>
<td>100</td>
<td>0.498137307078</td>
<td>0.484221417486</td>
<td>0.491080797482</td>
</tr>
<tr>
<td>150</td>
<td>0.538929440389</td>
<td>0.427812650893</td>
<td>0.476985195155</td>
</tr>
<tr>
<td rowspan="3">3</td>
<td>50</td>
<td>0.738186462324</td>
<td>0.544256120527</td>
<td>0.626558265583</td>
</tr>
<tr>
<td>100</td>
<td>0.667063020214</td>
<td>0.270752895753</td>
<td>0.385169927909</td>
</tr>
<tr>
<td>150</td>
<td>0.458050847458</td>
<td>0.426598263615</td>
<td>0.441765427054</td>
</tr>
<tr>
<td rowspan="3">4</td>
<td>50</td>
<td>0.507448789572</td>
<td>0.581023454158</td>
<td>0.541749502982</td>
</tr>
<tr>
<td>100</td>
<td>0.660091047041</td>
<td>0.26395631068</td>
<td>0.37711313394</td>
</tr>
<tr>
<td>150</td>
<td>0.711198428291</td>
<td>0.2145820984</td>
<td>0.329690346084</td>
</tr>
<tr>
<td rowspan="3">5</td>
<td>50</td>
<td>0.673548387097</td>
<td>0.509765625</td>
<td>0.580322401334</td>
</tr>
<tr>
<td>100</td>
<td>0.759791122715</td>
<td>0.178199632578</td>
<td>0.28869047619</td>
</tr>
<tr>
<td>150</td>
<td>0.518197573657</td>
<td>0.19606557377</td>
<td>0.28449096099</td>
</tr>
<tr>
<td rowspan="3">6</td>
<td>50</td>
<td>0.712121212121</td>
<td>0.488196411709</td>
<td>0.579271708683</td>
</tr>
<tr>
<td>100</td>
<td>0.571045576408</td>
<td>0.167189952904</td>
<td>0.258652094718</td>
</tr>
<tr>
<td>150</td>
<td>0.545180722892</td>
<td>0.209612044007</td>
<td>0.302802174822</td>
</tr>
 <tr>
<td rowspan="3">7</td>
<td>50</td>
<td>0.485947416138</td>
<td>0.619653179191</td>
<td>0.544715447154</td>
</tr>
<tr>
<td>100</td>
<td>0.563492063492</td>
<td>0.214016578749</td>
<td>0.310212998362</td>
</tr>
<tr>
<td>150</td>
<td>0.551656920078</td>
<td>0.16915720263</td>
<td>0.258920402562</td>
</tr>

<h3>Results for Pheonix:-</h3>
<table>
<tr>
<th>
Partition Number
</th>
<th>
Number of members
</th>
<th>
Precision
</th>
<th>
Recall
</th>
<th>
F-measure
</th>
</tr>
<tr>
<td rowspan="3">
1
</td>
<td>
50
</td>
<td>
0.48322147651
</td>
<td>
0.356435643564
</td>
<td>
0.410256410256
</td>
</tr>
<tr>
<td>
100
</td>
<td>
0.462082912032
</td>
<td>
0.536594911937
</td>
<td>
0.496559217675
</td>
</tr>
<tr>
<td>
150
</td>
<td>
0.417532640554
</td>
<td>
0.532630863358
</td>
<td>
0.468110530246
</td>
</tr>

<tr>
<td rowspan="3">
2
</td>
<td>
50
</td>
<td>
0.76376146789
</td>
<td>
0.256351039261
</td>
<td>
0.38386167147
</td>
</tr>
<tr>
<td>
100
</td>
<td>
0.499249249249
</td>
<td>
0.348167539267
</td>
<td>
0.410240592227
</td>
</tr>
<tr>
<td>
150
</td>
<td>
0.446632782719
</td>
<td>
0.307793345009
</td>
<td>
0.3644375324
</td>
</tr>

<tr>
<td rowspan="3">
3
</td>
<td>
50
</td>
<td>
0.512295081967
</td>
<td>
0.59694364852
</td>
<td>
0.551389501544
</td>
</tr>
<tr>
<td>
100
</td>
<td>
0.488441461596
</td>
<td>
0.467189728959
</td>
<td>
0.477579292745
</td>
</tr>
<tr>
<td>
150
</td>
<td>
0.507783145464
</td>
<td>
0.466929911155
</td>
<td>
0.486500385703
</td>
</tr>

<tr>
<td rowspan="3">
4
</td>
<td>
50
</td>
<td>
0.567890691716
</td>
<td>
0.546425636812
</td>
<td>
0.556951423786
</td>
</tr>
<tr>
<td>
100
</td>
<td>
0.471113758189
</td>
<td>
0.468047337278
</td>
<td>
0.469575541704
</td>
</tr>
<tr>
<td>
150
</td>
<td>
0.472628499791
</td>
<td>
0.50000000000
</td>
<td>
0.485929108485
</td>
</tr>

<tr>
<td rowspan="3">
5
</td>
<td>
50
</td>
<td>
0.614572333685
</td>
<td>
0.501291989664
</td>
<td>
0.552182163188
</td>
</tr>
<tr>
<td>
100
</td>
<td>
0.446226415094
</td>
<td>
0.521787093216
</td>
<td>
0.481057716756
</td>
</tr>
<tr>
<td>
150
</td>
<td>
0.496774193548
</td>
<td>
0.465364946536
</td>
<td>
0.480556889102
</td>
</tr>

<tr>
<td rowspan="3">
6
</td>
<td>
50
</td>
<td>
0.589261744966
</td>
<td>
0.612273361227
</td>
<td>
0.600547195622
</td>
</tr>
<tr>
<td>
100
</td>
<td>
0.549869904597
</td>
<td>
0.459087617668
</td>
<td>
0.500394632991
</td>
</tr>
<tr>
<td>
150
</td>
<td>
0.489247311828
</td>
<td>
0.362756264237
</td>
<td>
0.416612164814
</td>
</tr>

<tr>
<td rowspan="3">
7
</td>
<td>
50
</td>
<td>
0.653846153846
</td>
<td>
0.429420505201
</td>
<td>
0.518385650224
</td>
</tr>
<tr>
<td>
100
</td>
<td>
0.342857142857
</td>
<td>
0.506181818182
</td>
<td>
0.408810572687
</td>
</tr>
<tr>
<td>
150
</td>
<td>
0.399623588457
</td>
<td>
0.367358708189
</td>
<td>
0.3828125
</td>
</tr>
</table>

<h3>Results for San Jose:-</h3>
<table>
<tbody>
<tr><th>Partition Number</th><th>Number of members</th><th>Precision</th><th>Recall</th><th>F-measure</th></tr>
<tr>
<td rowspan="3">1</td>
<td>50</td>
<td>0.616191904048</td>
<td>0.568858131488</td>
<td>0.59157970493</td>
</tr>
<tr>
<td>100</td>
<td>0.517844136926</td>
<td>0.458413926499</td>
<td>0.486320109439</td>
</tr>
<tr>
<td>150</td>
<td>0.461823573017</td>
<td>0.332443970117</td>
<td>0.386596338815</td>
</tr>
<tr>
<td rowspan="3">
  2</td>
<td>50</td>
<td>0.605660377358</td>
<td>0.485138539043</td>
<td>0.538741258741</td>
</tr>
<tr>
<td>100</td>
<td>0.62962962963</td>
<td>0.296511627907</td>
<td>0.403162055336</td>
</tr>
<tr>
<td>150</td>
<td>0.598518518519</td>
<td>0.333884297521</td>
<td>0.428647214854</td>
</tr>
<tr>
<td rowspan="3">3</td>
<td>50</td>
<td>0.768229166667</td>
<td>0.431602048281</td>
<td>0.552693208431</td>
</tr>
<tr>
<td>100</td>
<td>0.705242334322</td>
<td>0.427458033573</td>
<td>0.532288167227</td>
</tr>
<tr>
<td>150</td>
<td>0.763870967742</td>
<td>0.304370179949</td>
<td>0.435294117647</td>
</tr>
<tr>
<td rowspan="3">4</td>
<td>50</td>
<td>0.743690851735</td>
<td>0.711698113208</td>
<td>0.727342846124</td>
</tr>
<tr>
<td>100</td>
<td>0.777908343126</td>
<td>0.433529796988</td>
<td>0.55677039529</td>
</tr>
<tr>
<td>150</td>
<td>0.572244897959</td>
<td>0.379534380076</td>
<td>0.456380208333</td>
</tr>
<tr>
<td rowspan="3">5</td>
<td>50</td>
<td>0.781136638452</td>
<td>0.622350674374</td>
<td>0.692761394102</td>
</tr>
<tr>
<td>100</td>
<td>0.755555555556</td>
<td>0.473607038123</td>
<td>0.582244254169</td>
</tr>
<tr>
<td>150</td>
<td>0.692610837438</td>
<td>0.407536231884</td>
<td>0.513138686131</td>
</tr>
<tr>
<td rowspan="3">6</td>
<td>50</td>
<td>0.580968280467</td>
<td>0.395005675369</td>
<td>0.47027027027</td>
</tr>
<tr>
<td>100</td>
<td>0.570502431118</td>
<td>0.338787295476</td>
<td>0.425120772947</td>
</tr>
<tr>
<td>150</td>
<td>0.555425904317</td>
<td>0.318821165439</td>
<td>0.405106382979</td>
</tr>
  <tr>
<td rowspan="3">7</td>
<td>50</td>
<td>0.591549295775</td>
<td>0.288858321871</td>
<td>0.388170055453</td>
</tr>
<tr>
<td>100</td>
<td>0.565723793677</td>
<td>0.357894736842</td>
<td>0.438426821406</td>
</tr>
<tr>
<td>150</td>
<td>0.551656920078</td>
<td>0.16915720263</td>
<td>0.258920402562</td>
</tr>
</tbody>
</table>

We also observed the importance of different feature vectors when used in Random Forest classifier. Below is the bar graph showing importance of each feature.

<img src="feature_importance.jpg">

As we can see that the content feature has the most weight. This means that content feature is the most important feature while performing the classification. Also, location of the event and user has a good significance. However, time and group frequency were found to be less important in classification. One of the main reasons why content feature has the most significance is because description of the event is highly rich and it shows the interest of the user. Location is also making a good impact because it recommends the user events based on the previous events he/she attended. 

## Conclusion and Next Steps
In sum, we propose a recommender system for event based social networks. We resolved the cold start problem associated with the event recommendations by using content based features: location, time, description and group frequency.

Further, the next steps would be to use data from other major sites like eventbrite.com, facebook.com and evaluate our recommendation on it. Another work could be to use more features for the system, like: Multi-relational Bayesian personalized ranking. We can also perform learning to rank on our recommender using Rank-SVM or coordinate ascent.

## References: 
[1] E. Minkov, B. Charrow, J. Ledlie, S. Teller, and T. Jaakkola, Collaborative future event recommendation, In Proc. of CIKM, pages 819-828, 2010.

[2] Augusto Q. Macedo, Leandro B. Marinho and Rodrygo L. T. Santos, Context-Aware Event Recommendation in Event based Social Networks, In Proc. of Recsys, pages 123-130, 2015.

[3] H. Khrouf and R. Troncy. Hybrid event recommendation using linked data and user diversity, In Proc. of RecSys, pages 185-192, 2013.

[4] A. Q. Macedo and L. B. Marinho, Event recommendation in event-based social networks, In Proc. of Int. Work. on Social Personalization, 2014.